ably/options.go

```diff
@@ -450,37 +474,62 @@ func (opts *clientOptions) activePort() (port int, isDefault bool) {
 	return
 }
 
-func (opts *clientOptions) getRestHost() string {
+// endpointFqdn handles an endpoint that uses a hostname, which may be an IPv4
+// address, IPv6 address or localhost
+func endpointFqdn(endpoint string) bool {
+	return strings.Contains(endpoint, ".") || strings.Contains(endpoint, "::") || endpoint == "localhost"
+}
+
+func (opts *clientOptions) endpointValueWithLegacySupport() string {
+	if !empty(opts.Endpoint) {
+		return opts.Endpoint
+	}
+
+	if !empty(opts.Environment) {
+		if opts.Environment == "production" {
+			return defaultOptions.Endpoint
+		}
+
+		return opts.Environment
+	}
+
 	if !empty(opts.RESTHost) {
 		return opts.RESTHost
 	}
-	if !opts.isProductionEnvironment() {
-		return opts.Environment + "-" + defaultOptions.RESTHost
-	}
-	return defaultOptions.RESTHost
-}
 
-func (opts *clientOptions) getRealtimeHost() string {
 	if !empty(opts.RealtimeHost) {
 		return opts.RealtimeHost
 	}
-	if !empty(opts.RESTHost) {
-		logger := opts.LogHandler
-		logger.Printf(LogWarning, "restHost is set to %s but realtimeHost is not set so setting realtimeHost to %s too. If this is not what you want, please set realtimeHost explicitly.", opts.RESTHost, opts.RealtimeHost)
-		return opts.RESTHost
+
+	return defaultOptions.Endpoint
+}
+
+// REC2
+func (opts *clientOptions) getEndpoint() string {
+	ep := opts.endpointValueWithLegacySupport()
+
+	if endpointFqdn(ep) {
+		return ep
 	}
-	if !opts.isProductionEnvironment() {
-		return opts.Environment + "-" + defaultOptions.RealtimeHost
+
+	if ep == "sandbox" {
+		return "sandbox.realtime.ably-nonprod.net"
```

Comment:
do we think having backcompat behaviour for "sandbox" is worth specifying so all sdks do it?

Reply:
When I work on other SDKs I intend to add the same behaviour, is this something that we would update the spec with? I have never done this.

Reply:
Yeah if we want all sdks to do it consistently it should be specified -- it's just a textile file that lives here https://github.com/ably/specification/blob/main/textile/features.textile?plain=1#L72-L103 , you can just do a pr to add a new spec item. I can if you like, but given you're making sdk prs probably good to do it so you know the flow